### DISCUSSION TITLE
Questions & Concerns about bOS Web Engine

### DISCUSSION THREAD

**Discussion by matiasberaldo** [near/react-on-chain#254] [Q&A] [answered]
Hello everyone!

I've been diving in the source code of the brand new bOS Web Engine and so far looks quite promising, however, some questions and concerns raised during the process, so I have opened this discussion to ask them, and also encourage anyone that has any question or concern regarding the new bOS Web Engine to add it here.

1. Currently we have more than 10k components running on the current Social VM, and many of them have specific logic that is adapted to run under the Social VM's limited environment. How is it planned to support these components and ensure retrocompatibility when bOS Web Engine gets released? I am concerned that if the process would require builders to migrate their existing components it will result in the lost of many valuable work, as many builders have hundred of components, and the effort and gas fees required to migrate them would be too high for the award.
2. We all agree that adding packages will add more functionalities to bOS, and therefore, will potentially increase the Developer Experience and attract more talent to the NEAR Ecosystem. However, having more functionalities also increases the amount of code per component, so when it comes to gas cost effectiveness, how is it planned to keep the gas costs affordable enough to not make it unattractive to develop a full app using bOS Web Engine? As we understand that one of the strongest points of the VM compared to deploying frontend apps to other services using IPFS, is the low amount of money required to pay to deploy components forever onchain.
3. The bOS Web Engine feature that enables loading NPM packages on demand is certainly awesome, however, I'm quite concerned about the centralization aspect that using public CDNs will inherently add to bOS components / apps. Are you planning to create a decentralized CDN service to serve NPM packages? Or, in other hand, use an external service that helps with that feature?

For now that's all, thank you for your time!

Best regards,
Matt

**Comment by mpeterdev** [comment, marked as answer]
Hi Matt, thanks for your thoughtful questions 🙌 

> Currently we have more than 10k components running on the current Social VM, and many of them have specific logic that is adapted to run under the Social VM's limited environment. How is it planned to support these components and ensure retrocompatibility when bOS Web Engine gets released? I am concerned that if the process would require builders to migrate their existing components it will result in the lost of many valuable work, as many builders have hundred of components, and the effort and gas fees required to migrate them would be too high for the award.

There are no plans for backwards compatibility, however we will be working hard to make migrations as painless as possible. We expect the effort and gas costs of migrations to be reasonable. By the nature of BOS, if a developer chooses not to upgrade their components for BOS Web Engine usage, then it can easily be forked and migrated by any developers who wish to build upon it.

> We all agree that adding packages will add more functionalities to bOS, and therefore, will potentially increase the Developer Experience and attract more talent to the NEAR Ecosystem. However, having more functionalities also increases the amount of code per component, so when it comes to gas cost effectiveness, how is it planned to keep the gas costs affordable enough to not make it unattractive to develop a full app using bOS Web Engine? As we understand that one of the strongest points of the VM compared to deploying frontend apps to other services using IPFS, is the low amount of money required to pay to deploy components forever onchain.

Since dependencies are loaded in the background instead of bundled into component code, they will not affect component size. The change which would likely most increase the size of source code under BWE would be the ability to write components in TypeScript. This is optional, if a developer wishes to fully prioritize source code size then they can continue writing in JS. For most, I think the code quality that TypeScript affords will outweigh the increased storage cost.

> The bOS Web Engine feature that enables loading NPM packages on demand is certainly awesome, however, I'm quite concerned about the centralization aspect that using public CDNs will inherently add to bOS components / apps. Are you planning to create a decentralized CDN service to serve NPM packages? Or, in other hand, use an external service that helps with that feature?

This is a good callout, and something touched on in our _[Press Release/ FAQ](https://github.com/near/bos-web-engine/blob/main/press-release-faq.md#how-are-off-chain-dependencies-decentralized)_ document. Copying those contents here:

> **How are off-chain dependencies decentralized?**
> Packages are loaded via CDNs, and most CDNs have decentralization architecture built in. That being said, it is possible for a CDN to go offline or be blocked in a particular region. In this case, the dependency will fail to load which may cause the component to fail to render.
> 
> In the spirit of progressive decentralization, BWE will launch with dependencies being powered directly by CDN URLs. In the future, we will explore the following:
> 
> - a package registry hosted on an inherently decentralized network (e.g. IPFS)
> - creating CDN-agnostic import syntax where developers can specify the package and version they require then leave resolution up to the gateway at runtime

We're better served by moving quickly towards production availability and then expanding our decentralization architecture once we've achieved adoption.

**Reply by matiasberaldo** [reply]
Hello!

Thank you for your detailed reply. Now it is more clear and I feel pretty confident that this hard work you are doing will take NEAR ecosystem to explore new horizons and possibilities.

Thanks again,
Matt
